src/workerd/api/global-scope.c++

```diff
@@ -706,7 +706,20 @@ void ServiceWorkerGlobalScope::queueMicrotask(jsg::Lock& js, v8::Local<v8::Funct
               return jsg::check(
                   function->Call(js.v8Context(), js.v8Null(), argv.size(), argv.data()));
             }, [&](jsg::Value exception) {
-              reportError(js, jsg::JsValue(exception.getHandle(js)));
+              // The reportError call itself can potentially throw errors. Let's catch
+              // and report them as well.
+              js.tryCatch([&] { reportError(js, jsg::JsValue(exception.getHandle(js))); },
+                  [&](jsg::Value exception) {
+                if (IoContext::hasCurrent()) {
+                // If we're running in a request context then we want to abort the
+                // request with the exception here.
+                IoContext::current().abort(js.exceptionToKj(kj::mv(exception)));
```

Comment:
Is there some background on why this is considered the right thing to do?

Reply:
If an error is thrown in the error event handler it was getting printed but otherwise would continue as if nothing happened. For Node.js compat, however, which uses queueMicrotask as the underlying impl of process.nextTick we need roughly the equivalent of a process crash, the closest we have and would want to get is erroring the request. Otherwise it just becomes a foot gun. @anonrig discovered the inconsistency while trying to port node.js tests for zlib over.

Reply:
I forget, is reportError() a node compat API or does it exist in browsers? In browsers, what happens if it throws?

I do agree that node APIs should emulate node behavior and the best way to emulate a process crash would be to abort the I/O context. But I don't agree that "crash on error" is _inherently_ the right thing to do.

Reply:
`reportError(...)` is a standard api.

![image](https://github.com/user-attachments/assets/2d9c8201-6357-46de-b095-c36c822ef4a6)

Reply:
So in the browser, the event handler throwing doesn't "abort" anything. It feels wrong, then, for this event handler to work differently on Workers.

Can we limit this aborting behavior to happen specifically when the microtask was queued using `process.nextTick`?

Or maybe apps should just register a callback themselves that calls `ctx.abort()` if that's what they want?

Reply:
The thing that worries me here is that Node in general has a policy of "abort on uncaught errors", but browsers do not, and Workers have always behaved more like browsers. It's weird to have one single place where we actually do abort on uncaught exceptions suddenly. This doesn't feel consistent.

I'm all for apps being able to opt into it if they want it but the defaults should be consistent I think?

Reply:
I generally disagree but I can update this to only about if the nodejs-compat flag is enabled. That way it is explicitly opt in.

Reply:
But that's still inconsistent? Doesn't node also abort on unhandled rejections and such? Also I feel like node-compat should just make node APIs available, but not change the behavior of non-node APIs.

I could imagine a _separate_ compat flag which opts into "eagerly abort on errors", which would apply both here and to unhandled rejections.